### PR TITLE
Add support to mount config maps and secrets to mgw deployment

### PR DESCRIPTION
**Pull request by Jayanie**
## Purpose

- Add support to mount config maps and secrets to micro-gateway deployment.

- User can add **mgw-deployment-configs** to controller_conf.yaml and provide the config maps and secrets to be added to micro-gateway deployment.
- The config maps and secrets to be added need to be in the same namespace which the **mgw-deployment-configs** is being deployed.

> Approach 02

- User can create and deploy a **mgw-deployment-configs** config map in the same namespace which the API is being deployed and in there he/she can provide the config maps and secrets to be added to micro-gateway deployment.
- The config maps and secrets to be added need to be in the same namespace which the **mgw-deployment-configs**  is being deployed.

### CODE REVIEW

File: api-operator/pkg/mgw/deployment.go [hunk 1]
```diff
@@ -62,11 +84,36 @@ func AddContainers(containers *[]corev1.Container) {
...
+	mgwDeploymentConfMap := k8s.NewConfMap()
+	errGetDeploy := k8s.Get(client, types.NamespacedName{Name: mgwDeploymentConfigMapName, Namespace: api.Namespace}, mgwDeploymentConfMap)
+	if errGetDeploy != nil && errors.IsNotFound(errGetDeploy) {
+		logDeploy.Info("Get mgw deployment configs", "from namespace", wso2NameSpaceConst)
+		//retrieve mgw deployment configs from wso2-system namespace
+		err := k8s.Get(client, types.NamespacedName{Namespace: wso2NameSpaceConst, Name: mgwDeploymentConfigMapName},
+			mgwDeploymentConfMap)
+		if err != nil {
+			logDeploy.Error(err, "MGW Deployment configs not defined")
+			return nil
```
Comment (renuka-fernando) [inline api-operator/pkg/mgw/deployment.go:104]: Shouldn't we return error if it is not a `errors.IsNotFound(err)`? if it is not found can't we continue without returning.
```suggestion
			if !errors.IsNotFound(err){
			    return err
			}
			// else continue without returning
```

File: api-operator/pkg/mgw/deployment.go [hunk 2]
```diff
@@ -62,11 +84,36 @@ func AddContainers(containers *[]corev1.Container) {
...
+		}
+	} else if errGetDeploy != nil {
+		logDeploy.Error(errGetDeploy, "Error getting mgw deployment configs from user namespace")
+	} else {
+		//retrieve mgw deployment configs from user namespace
+		err := k8s.Get(client, types.NamespacedName{Namespace: api.Namespace, Name: mgwDeploymentConfigMapName},
+			mgwDeploymentConfMap)
```
Comment (renuka-fernando) [inline api-operator/pkg/mgw/deployment.go:111]: Shouldn't we have already `k8s.Get` this config map above? shall we delete this whole `else` block.

File: api-operator/pkg/mgw/deployment.go [hunk 3]
```diff
@@ -85,6 +132,74 @@ func Deployment(api *wso2v1alpha1.API, controlConfigData map[string]string, owne
 		deployVolume = append(deployVolume, *analVol)
 		deployVolumeMount = append(deployVolumeMount, *analMount)
 	}
+
+	var deploymentConfigMaps []DeploymentConfig
+	yamlErrDeploymentConfigMaps := yaml.Unmarshal([]byte(mgwDeploymentConfMap.Data[mgwConfigMaps]), &deploymentConfigMaps)
+	if yamlErrDeploymentConfigMaps != nil {
+		logHpa.Error(yamlErrDeploymentConfigMaps, "Error marshalling mgw config maps yaml",
```
Comment (renuka-fernando) [inline api-operator/pkg/mgw/deployment.go:139]:
```suggestion
		logDeploy.Error(yamlErrDeploymentConfigMaps, "Error marshalling mgw config maps yaml",
```

File: api-operator/pkg/mgw/deployment.go [hunk 4]
```diff
@@ -85,6 +132,74 @@ func Deployment(api *wso2v1alpha1.API, controlConfigData map[string]string, owne
...
+			"configmap", mgwDeploymentConfMap)
+	}
+	var deploymentSecrets []DeploymentConfig
+	yamlErrDeploymentSecrets := yaml.Unmarshal([]byte(mgwDeploymentConfMap.Data[mgwSecrets]), &deploymentSecrets)
+	if yamlErrDeploymentSecrets != nil {
+		logHpa.Error(yamlErrDeploymentSecrets, "Error marshalling mgw secrets yaml", "configmap",
```
Comment (renuka-fernando) [inline api-operator/pkg/mgw/deployment.go:145]:
```suggestion
		logDeploy.Error(yamlErrDeploymentSecrets, "Error marshalling mgw secrets yaml", "configmap",
```

**Comment by renuka-fernando** [comment]
This is a suggestion; shall we move this logic to a new file? `customvolumes.go` or `uservolumes.go` in mgw package.
@pubudu538

File: api-operator/pkg/mgw/deployment.go [hunk 5]
```diff
@@ -48,6 +54,22 @@ const (
...
+)
+
+type DeploymentConfig struct {
+	Name          string `yaml:"name"`
+	MountLocation string `yaml:"mountLocation"`
+	SubPath       string `yaml:"subPath"`
+	Namespace     string `yaml:"namespace,omitempty"`
+}
+
+var logDeploy = log.Log.WithName("mgw.hpa")
```
[the remark this answers is not in the document]
Comment (Jayanie) [reply api-operator/pkg/mgw/deployment.go]: Fixed with 533f5d267babd0d26a517e9eaff29e76d9a34e91

File: api-operator/pkg/mgw/deployment.go (hunk 1 above)
Comment (Jayanie) [reply api-operator/pkg/mgw/deployment.go]: Fixed with 533f5d267babd0d26a517e9eaff29e76d9a34e91

File: api-operator/pkg/mgw/deployment.go (hunk 2 above)
Comment (Jayanie) [reply api-operator/pkg/mgw/deployment.go]: Fixed with 533f5d267babd0d26a517e9eaff29e76d9a34e91

File: api-operator/pkg/mgw/deployment.go (hunk 3 above)
Comment (Jayanie) [reply api-operator/pkg/mgw/deployment.go]: Fixed with 533f5d267babd0d26a517e9eaff29e76d9a34e91

File: api-operator/pkg/mgw/deployment.go (hunk 4 above)
Comment (Jayanie) [reply api-operator/pkg/mgw/deployment.go]: Fixed with 533f5d267babd0d26a517e9eaff29e76d9a34e91

File: api-operator/pkg/mgw/uservolumes.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	mgwDeploymentConfMap := k8s.NewConfMap()
+	errGetDeploy := k8s.Get(client, types.NamespacedName{Name: mgwDeploymentConfigMapName, Namespace: api.Namespace},
+		mgwDeploymentConfMap)
+	if errGetDeploy != nil && errors.IsNotFound(errGetDeploy) {
+		logDeploy.Info("Get mgw deployment configs", "from namespace", wso2NameSpaceConst)
+		//retrieve mgw deployment configs from wso2-system namespace
+		err := k8s.Get(client, types.NamespacedName{Namespace: wso2NameSpaceConst, Name: mgwDeploymentConfigMapName},
+			mgwDeploymentConfMap)
+		if err != nil && !errors.IsNotFound(err) {
+			logDeploy.Error(err, "MGW Deployment configs not defined")
```
Comment (renuka-fernando) [inline api-operator/pkg/mgw/uservolumes.go:65]: The log should be "Error while reading user volumes config" intead of "Not found"

**Pull request merged**